### ISSUE TITLE
InvalidArgumentsError With signTypedData

### ISSUE THREAD

**Issue by charlie-kim** [NomicFoundation/edr#346]
### Version of Hardhat

2.22.1

### What happened?

I have a hardhat test with `signTypedData` that works in `2.19.5` but not in `2.22.1`. Something to do with `EdrProviderWrapper`?

I was redirected from this issue: https://github.com/pcaversaccio/xdeployer/issues/245

### Minimal reproduction steps

```ts
import { ethers } from "hardhat";

const [signer] = await ethers.getSigners();
const domain = {
  name: 'Collateral',
  version: '2',
  chainId: 31337, // hardhat
  verifyingContract: '0xb0279Db6a2F1E01fbC8483FCCef0Be2bC6299cC3',
  salt: new Uint8Array([
    189, 181,  85, 137, 189, 110, 138, 202,
     85,  30, 222, 158, 181, 116,  29,  49,
    117,  28,  64, 185, 106, 101, 110, 218,
     99,  63,  69, 230, 136, 254,  92, 224
  ])
};
const type = {
  UpdateAdminThreshold: [
    { name: 'threshold', type: 'uint8' },
    { name: 'nonce', type: 'uint256' }
  ]
} ;
const data = { threshold: 1, nonce: 0n };

await signer.signTypedData(domain, type, data);
```

```
InvalidArgumentsError: The message parameter is an invalid JSON. at line 1 column 776
      at EdrProviderWrapper.request (/Users/charlie/Code/rain-contract-v2/node_modules/hardhat/src/internal/hardhat-network/provider/provider.ts:423:10)

response.error looks like this.


{
  code: -32602,
  message: 'The message parameter is an invalid JSON. at line 1 column 776',
  data: {
    method: 'eth_signTypedData_v4',
    params: [
      '0x976ea74026e726554db657fa54763abd0c3a0aa9',
      '{"types":{"UpdateAdminThreshold":[{"name":"threshold","type":"uint8"},{"name":"nonce","type":"uint256"}],"EIP712Domain":[{"name":"name","type":"string"},{"name":"version","type":"string"},{"name":"chainId","type":"uint256"},{"name":"verifyingContract","type":"address"},{"name":"salt","type":"bytes32"}]},"domain":{"name":"Collateral","version":"2","chainId":"0x7a69","verifyingContract":"0xb0279db6a2f1e01fbc8483fccef0be2bc6299cc3","salt":"0x54c6b2b3ad37d2ee0bf85cf73d4c147b0a1c333627a2cbf9a1bb9ecc1543fc7a"},"primaryType":"UpdateAdminThreshold","message":{"threshold":"1","nonce":"0"}}'
    ]
  }
}
```

### Search terms

EdrProviderWrapper signTypedData InvalidArgumentsError

**Comment by agostbiro** [comment]
Thanks for the detailed bug report @charlie-kim ! I found the issue. Unfortunately it looks like a bug in a dependency so we have to fix it there first, get the fix released and then upgrade the dependency which might take some time.

As a workaround, you could replace the salt hexadecimal string with an array of u8 values, e.g. this works:

```json
{
  "types": {
    "UpdateAdminThreshold": [
      {
        "name": "threshold",
        "type": "uint8"
      },
      {
        "name": "nonce",
        "type": "uint256"
      }
    ],
    "EIP712Domain": [
      {
        "name": "name",
        "type": "string"
      },
      {
        "name": "version",
        "type": "string"
      },
      {
        "name": "chainId",
        "type": "uint256"
      },
      {
        "name": "verifyingContract",
        "type": "address"
      },
      {
        "name": "salt",
        "type": "bytes32"
      }
    ]
  },
  "domain": {
    "name": "Collateral",
    "version": "2",
    "chainId": "0x7a69",
    "verifyingContract": "0xb0279db6a2f1e01fbc8483fccef0be2bc6299cc3",
   // Note: replaced "0x54c6b2b3ad37d2ee0bf85cf73d4c147b0a1c333627a2cbf9a1bb9ecc1543fc7a" with the array for the salt
    "salt": [
      84,
      198,
      178,
      179,
      173,
      55,
      210,
      238,
      11,
      248,
      92,
      247,
      61,
      76,
      20,
      123,
      10,
      28,
      51,
      54,
      39,
      162,
      203,
      249,
      161,
      187,
      158,
      204,
      21,
      67,
      252,
      122
    ]
  },
  "primaryType": "UpdateAdminThreshold",
  "message": {
    "threshold": "1",
    "nonce": "0"
  }
}
```

**Comment by charlie-kim** [comment]
Hello @agostbiro , thank you for your response.

Unfortunately, I am getting a typescript error with `signTypedData` when I try to use number array.

```
Argument of type '{ name: string; version: string; chainId: number; verifyingContract: string; salt: number[]; }' is not assignable to parameter of type 'TypedDataDomain'.
  Types of property 'salt' are incompatible.
    Type 'number[]' is not assignable to type 'BytesLike | null | undefined'.
      Type 'number[]' is missing the following properties from type 'Uint8Array': BYTES_PER_ELEMENT, buffer, byteLength, byteOffset, and 3 more.ts(2345)
```

**Comment by humanshield89** [comment]
Can confirm this. my workaround is I downgraded hardhat to `2.19.5`, because unfortunately sending the array of u8 still gives errors

**Comment by agostbiro** [comment]
Thanks for the info @charlie-kim and @humanshield89 and apologies for the confusion regarding the proposed work around: I only tested with Rust, not TypeScript. 

We've released [EDR v0.3.3](https://github.com/NomicFoundation/hardhat/blob/f0b18441bc4a482c37026b7d4a56783d5ce0a749/crates/edr_napi/CHANGELOG.md#033) now to support hex string as salt in `eth_signTypedData_v4`. In order to use EDR v0.3.3, if you're using the latest Hardhat release, you can remove your `node_modules` and lock file and reinstall your dependencies. Otherwise you can wait for the next Hardhat release which will automatically use EDR v0.3.3.

**Comment by charlie-kim** [comment]
Confirming it works. Thanks for the fix, @agostbiro !
